diff --git a/configure.ac b/configure.ac
index 52bb7c7..7387fa2 100644
--- a/configure.ac
+++ b/configure.ac
@@ -761,6 +761,9 @@ case "${canonical}" in
   powerpc-ibm-aix[5-9]* | powerpc-ibm-aix[1-9][0-9]* )
     opsys=aix4-2
   ;;
+  i370-ibm-openedition)
+    opsys=zos
+  ;;
 
   ## Solaris
   *-*-solaris* | *-*-sunos*)
@@ -4430,7 +4433,7 @@ if test "$ac_cv_func_pthread_setname_np" = "yes"; then
 fi
 
 dnl No need to check for posix_memalign if aligned_alloc works.
-AC_CHECK_FUNCS([aligned_alloc posix_memalign], [break])
+AC_CHECK_FUNCS([aligned_alloc], [break])
 AC_CHECK_DECLS([aligned_alloc], [], [], [[#include <stdlib.h>]])
 
 case $with_unexec,$canonical in
@@ -4845,9 +4848,9 @@ fi
 dnl Everybody supports this, except MS.
 dnl Seems like the kind of thing we should be testing for, though.
 ## Note: PTYs are broken on darwin <6.  Use at your own risk.
-if test "${opsys}" != "mingw32"; then
-  AC_DEFINE(HAVE_PTYS, 1, [Define if the system supports pty devices.])
-fi
+#if test "${opsys}" != "mingw32"; then
+#  AC_DEFINE(HAVE_PTYS, 1, [Define if the system supports pty devices.])
+#fi
 
 dnl Everybody supports this, except MS-DOS.
 dnl Seems like the kind of thing we should be testing for, though.
