diff --git a/src/editfns.c b/src/editfns.c
index 203d3a4..e32cb4e 100644
--- a/src/editfns.c
+++ b/src/editfns.c
@@ -1266,11 +1266,15 @@ is in general a comma-separated list.  */)
   if (!pw)
     return Qnil;
 
+/*FIXME: Invesigate USER_FULL_NAME */
+#ifndef __MVS__
   p = USER_FULL_NAME;
+  
   /* Chop off everything after the first comma, since 'pw_gecos' is a
      comma-separated list. */
   q = strchr (p, ',');
   full = make_string (p, q ? q - p : strlen (p));
+#endif
 
 #ifdef AMPERSAND_FULL_NAME
   p = SSDATA (full);
