diff --git a/lib-src/emacsclient.c b/lib-src/emacsclient.c
index 57a5eff..99e1129 100644
--- a/lib-src/emacsclient.c
+++ b/lib-src/emacsclient.c
@@ -1412,8 +1412,14 @@ local_sockname (int s, char sockname[socknamesize], int tmpdirlen,
   char *emacsdirend = sockname + tmpdirlen + suffixlen -
     strlen(server_name) - 1;
   *emacsdirend = '\0';
+/*FIXME: Investigate adding openat */
+#ifdef __MVS__
+  int dir = open (sockname,
+		    O_PATH );
+#else
   int dir = openat (AT_FDCWD, sockname,
 		    O_PATH | O_DIRECTORY | O_NOFOLLOW | O_CLOEXEC);
+#endif
   *emacsdirend = '/';
   if (dir < 0)
     return errno;
